File: agent/packet_engine/go/syn_detector.go
package main

import (
    "fmt"
    "os/exec"
    "sync"
    "time"
)

// Simple SYN flood detector skeleton.
// Counts SYN and ACKs per source IP in 1-second windows.

type synStats struct {
    synTimestamps []time.Time
    ackTimestamps []time.Time
}

var (
    mu      sync.Mutex
    stats   = make(map[string]*synStats)
)

// RecordPacket is called by packet capture when a SYN or ACK is seen.
func RecordPacket(srcIP string, isSYN bool, isACK bool) {
    mu.Lock()
    defer mu.Unlock()
    now := time.Now()
    s, ok := stats[srcIP]
    if !ok {
        s = &synStats{}
        stats[srcIP] = s
    }
    if isSYN {
        s.synTimestamps = append(s.synTimestamps, now)
    }
    if isACK {
        s.ackTimestamps = append(s.ackTimestamps, now)
    }
}

// DetectSYNFlood scans stats and triggers alert if SYN:ACK ratio > 10 in 1s window.
func DetectSYNFlood() {
    mu.Lock()
    defer mu.Unlock()
    cutoff := time.Now().Add(-1 * time.Second)
    for ip, s := range stats {
        // trim old
        synCount := 0
        ackCount := 0
        newSyn := s.synTimestamps[:0]
        for _, t := range s.synTimestamps {
            if t.After(cutoff) {
                synCount++
                newSyn = append(newSyn, t)
            }
        }
        s.synTimestamps = newSyn

        newAck := s.ackTimestamps[:0]
        for _, t := range s.ackTimestamps {
            if t.After(cutoff) {
                ackCount++
                newAck = append(newAck, t)
            }
        }
        s.ackTimestamps = newAck

        ratio := 0.0
        if ackCount > 0 {
            ratio = float64(synCount) / float64(ackCount)
        } else if synCount > 0 {
            ratio = float64(synCount)
        }

        if synCount >= 20 && ratio > 10.0 {
            // Trigger an alert to AI Engine and log
            fmt.Printf("SYN flood suspected from %s: syn=%d ack=%d ratio=%.1f\n", ip, synCount, ackCount, ratio)
            // notify ML engine with basic metrics
            notifyML(ip)
            // Example action: enable SYN cookies
            _ = EnableTCPSynCookies()
        }
    }
}

// EnableTCPSynCookies uses system call to enable kernel SYN cookies (requires privileges).
func EnableTCPSynCookies() error {
    // sysctl -w net.ipv4.tcp_syncookies=1
    cmd := exec.Command("sysctl", "-w", "net.ipv4.tcp_syncookies=1")
    out, err := cmd.CombinedOutput()
    if err != nil {
        return fmt.Errorf("failed enable syncookies: %v output=%s", err, string(out))
    }
    fmt.Printf("Enabled tcp_syncookies: %s", string(out))
    return nil
}

// Background loop to periodically run detection.
func init() {
    go func() {
        ticker := time.NewTicker(1 * time.Second)
        for range ticker.C {
            DetectSYNFlood()
        }
    }()
}

File: agent/packet_engine/python/syn_detector.py
import time
import subprocess
from collections import defaultdict, deque

class SynDetector:
    def __init__(self, window_seconds=1):
        self.window = window_seconds
        # per-ip deque of (timestamp, type) where type is 'SYN' or 'ACK'
        self.data = defaultdict(lambda: deque())

    def record_packet(self, src_ip, is_syn, is_ack):
        now = time.time()
        if is_syn:
            self.data[src_ip].append((now, 'SYN'))
        if is_ack:
            self.data[src_ip].append((now, 'ACK'))

    def detect(self):
        cutoff = time.time() - self.window
        alerts = []
        for ip, dq in list(self.data.items()):
            # drop old
            while dq and dq[0][0] < cutoff:
                dq.popleft()
            syn_count = sum(1 for _, t in dq if t == 'SYN')
            ack_count = sum(1 for _, t in dq if t == 'ACK')
            ratio = float(syn_count) / (ack_count or 1)
            if syn_count >= 20 and (ack_count == 0 or ratio > 10.0):
                alerts.append((ip, syn_count, ack_count, ratio))
                # Example mitigation: enable syncookies
                self.enable_tcp_syncookies()
        return alerts

    def enable_tcp_syncookies(self):
        try:
            subprocess.run(['sysctl', '-w', 'net.ipv4.tcp_syncookies=1'], check=True)
        except Exception as e:
            print('Failed to enable syncookies:', e)
